mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala

```diff
@@ -1009,10 +1009,24 @@ private[spark] object RandomForest extends Logging {
       // sort distinct values
       val valueCounts = valueCountMap.toSeq.sortBy(_._1).toArray
 
+      def weightedMean(pre: (Double, Int), cru: (Double, Int)): Double = {
+        val (preValue, preCount) = pre
+        val (curValue, curCount) = cru
+        (preValue * preCount + curValue * curCount) / (preCount + curCount)
+      }
+
       // if possible splits is not enough or just enough, just return all possible splits
       val possibleSplits = valueCounts.length - 1
-      if (possibleSplits <= numSplits) {
-        valueCounts.map(_._1).init
+      if (possibleSplits == 0) {
+        // constant feature
+        Array.empty[Double]
+
+      } else if (possibleSplits <= numSplits) {
+        valueCounts
+          .sliding(2)
+          .map{x => weightedMean(x(0), x(1))}
```

Comment:
Nit: use () instead of {}
There are more efficient ways of writing this but not as compact. I think it's OK unless someone suggests this is performance critical here